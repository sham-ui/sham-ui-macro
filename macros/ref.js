const t = require( 'babel-types' );

/**
 * Macro for get/generate uniq reference
 * @function ref
 * @param {string} [name] Name of reference
 *
 * @example
 * // Config in package.json
 * // "babelMacros": {
 * //   "ref": {
 * //     "enabled": true,
 * //     "uniq": true
 * //   }
 * // }
 *
 * import { ref } from 'sham-ui-macro/ref.macro';
 *
 * const firstName = ref();
 * const lastName = ref();
 * const state = {
 *     [ firstName ]: 'John',
 *     [ lastName ]: 'Smith'
 * }
 * const fullName = state[ ref( 'firstName' ) ] + state[ ref( 'lastName' ) ];
 *
 * // ↓ ↓ ↓ ↓ ↓ ↓
 *
 * const firstName = 0;
 * const lastName = 1;
 * const state = {
 *     [ firstName ]: 'John',
 *     [ lastName ]: 'Smith'
 * }
 * const fullName = state[ 0 ] + state[ 1 ];
 */

/**
 * @param {BabelPath} callExpressionPath
 * @param {Object} config
 * @param {Object} state
 * @private
 */
module.exports = function ref( callExpressionPath, config, state ) {
    const { uniq = false } = config;

    let name;
    const args = callExpressionPath.node.arguments;
    if ( 0 === args.length  ) {
        const parentPath = callExpressionPath.parentPath;
        if ( !parentPath.isVariableDeclarator() ) {
            throw new Error( 'Pass name to ref' );
        }

        // Use variable as name
        name = parentPath.node.id.name;
    } else {
        name = args[ 0 ].value;
    }

    let identifier;
    if ( uniq ) {
        if ( !state.map.has( name ) ) {
            state.map.set(
                name,
                state.map.size
            );
        }
        identifier = t.NumericLiteral( state.map.get( name ) );
    } else {
        identifier = t.StringLiteral( name );
    }

    callExpressionPath.replaceWith(
        identifier
    );
};
