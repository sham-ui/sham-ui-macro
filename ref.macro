/* eslint-env node */
'use strict';
const { createMacro } = require( 'babel-macros' );
const ref = require( './macros/ref' );

const refState = {
    map: new Map()
};

module.exports = createMacro( macro, {
    configName: 'ref'
} );

function macro( { references, config } ) {
    const { enabled = false } = config;
    if ( enabled ) {

        // After transpile with rollup in cjs import will:
        // var ref_macro = require( 'sham-ui/ref.macro' );
        // ...
        // const a = ref_macro.ref();
        // ...
        // But for ES modules transpile to
        // import { ref } from 'sham-ui/ref.macro';
        ( references.default || references.ref ).forEach( referencePath => {
            ref(
                referencePath.findParent( x => x.isCallExpression() ),
                config,
                refState
            )
        } );
    } else {

        // Keep imports for process in app level
        return {
            keepImports: true
        }
    }
}
