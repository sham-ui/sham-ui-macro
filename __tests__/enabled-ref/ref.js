/* eslint-env node */

const path = require( 'path' );
const pluginTester = require( 'babel-plugin-tester' );
const plugin = require( 'babel-plugin-macros' );

const projectRoot = path.join( __dirname, '../../../' );

expect.addSnapshotSerializer( {
    print( val ) {
        return val.split( projectRoot ).join( '<PROJECT_ROOT>/' );
    },
    test( val ) {
        return typeof val === 'string';
    }
} );

pluginTester( {
    plugin,
    snapshot: true,
    babelOptions: {
        filename: __filename,
        parserOpts: {
            plugins: [
                'decorators',
                'classProperties'
            ]
        }
    },
    tests: [ {
        title: 'without args',
        code: `

            import { $ } from '../../ref.macro';

            const foo = $();
        `
    }, {
        title: 'with args',
        code: `
            import { $ } from '../../ref.macro';

            const foo = $( 'bar' ) ;
        `
    }, {
        title: 'as key',
        code: `
            import { $ } from '../../ref.macro';

            const state = { [ $( 'foo' ) ]: true };
        `
    }, {
        title: 'as member',
        code: `
            import { $ } from '../../ref.macro';

            const foo = $.bar ;
        `
    }, {
        title: 'member as key',
        code: `
            import { $ } from '../../ref.macro';

            const state = { [ $.foo ]: true };
        `
    }, {
        title: 'cjs without args',
        code: `

            import babel_macro from '../../ref.macro';

            const foo = babel_macro.$();
        `
    }, {
        title: 'cjs with args',
        code: `
            import babel_macro from '../../ref.macro';

            const foo = babel_macro.$( 'bar' ) ;
        `
    }, {
        title: 'cjs as key',
        code: `
            import babel_macro from '../../ref.macro';

            const state = { [ babel_macro.$( 'foo' ) ]: true };
        `
    }, {
        title: 'cjs as member',
        code: `
            import babel_macro from '../../ref.macro';

            const foo = babel_macro.$.bar ;
        `
    }, {
        title: 'cjs member as key',
        code: `
            import babel_macro from '../../ref.macro';

            const state = { [ babel_macro.$.foo ]: true };
        `
    }, {
        title: 'this$ assign',
        code: `
            import { this$ } from '../../ref.macro';

            this$.foo = 1;
        `
    }, {
        title: 'this$ call',
        code: `
            import { this$ } from '../../ref.macro';

            this$.foo( 1 );
        `
    }, {
        title: 'cjs this$ assign',
        code: `
            import babel_macro from '../../ref.macro';

            babel_macro.this$.foo = 1;
        `
    }, {
        title: 'cjs this$ call',
        code: `
            import babel_macro from '../../ref.macro';

            babel_macro.this$.foo( 1 );
        `
    }, {
        title: 'don\'t used',
        code: `
            import '../../ref.macro';
        `
    }, {
        title: '$ & this$',
        code: `
        import { this$, $ } from '../../ref.macro';

        function App( options, update ) {
            const firstGroup = $();
    
            const state = options( {
                [ firstGroup ]: true,
            } );
    
            this$.toggleGroup = () => update( {
                [ firstGroup ]: !state[ firstGroup ]
            } );
        }
        `
    } ]
} );
