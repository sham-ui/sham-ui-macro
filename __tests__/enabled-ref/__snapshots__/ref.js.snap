// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros as key: as key 1`] = `

import { ref } from '../../babel.macro';

const state = { [ ref( 'foo' ) ]: true };

      ↓ ↓ ↓ ↓ ↓ ↓

const state = { ['foo']: true };

`;

exports[`macros with args: with args 1`] = `

import { ref } from '../../babel.macro';

const foo = ref( 'bar' ) ;

      ↓ ↓ ↓ ↓ ↓ ↓

const foo = 'bar';

`;

exports[`macros without args: without args 1`] = `

import { ref } from '../../babel.macro';

const foo = ref();

      ↓ ↓ ↓ ↓ ↓ ↓

const foo = 'foo';

`;
